common/model/transfer.c

```diff
@@ -1,676 +1,474 @@
+                                   flex_trit_t const* const tag,
+                                   flex_trit_t const* const address,
+                                   int64_t value, uint64_t timestamp) {
+  transfer_value_out_t* value_out = NULL;
+  transfer_t* tf = NULL;
+  if (!validate_output(output)) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] output is invalid.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!address) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] address cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!tag) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] tag cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  tf = (transfer_t*)calloc(1, sizeof(transfer_t));
+  if (!tf) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  value_out = (transfer_value_out_t*)calloc(1, sizeof(transfer_value_out_t));
+  if (!value_out) {
+    free(tf);
+    return NULL;
+  }
+  memcpy(value_out, output, sizeof(transfer_value_out_t));
+
+  tf->type = VALUE_OUT;
+  tf->address = address;
+  tf->value = value;
+  tf->tag = tag;
+  tf->timestamp = timestamp;
+  tf->meta = (void*)value_out;
+  return tf;
+}
+
+// create a value_int transfer object.
+transfer_t* transfer_value_in_new(flex_trit_t const* const address,
+                                  flex_trit_t const* const tag, int64_t value,
+                                  flex_trit_t const* const data,
+                                  size_t data_len, uint64_t timestamp) {
+  transfer_value_in_t* value_in = NULL;
+  transfer_t* tf = NULL;
+  if (!address) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] address cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!tag) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] tag cannot be NULL.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  if (!data && data_len != 0) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] data length should be 0.\n",
+              __func__, __LINE__);
+    return NULL;
+  }
+
+  tf = (transfer_t*)calloc(1, sizeof(transfer_t));
+  if (!tf) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+
+  value_in = (transfer_value_in_t*)calloc(1, sizeof(transfer_value_in_t));
+  if (!value_in) {
+    free(tf);
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  value_in->len = data_len;
+  value_in->data = data;
+
+  tf->type = VALUE_IN;
+  tf->address = address;
+  tf->tag = tag;
+  tf->value = value;
+  tf->timestamp = timestamp;
+  tf->meta = (void*)value_in;
+  return tf;
+}
+
+void transfer_free(transfer_t** transfer) {
+  if (transfer || *transfer) {
+    free((*transfer)->meta);
+    free(*transfer);
+    *transfer = NULL;
+  }
```

Comment:
We should probably wipe out the `seed` field if the transfer is a `VALUE_OUT`

Reply:
that is a pointer in `transfer_value_out_t`

Reply:
Fair enough